alphastats/tl/differential_expression_analysis.py

```diff
@@ -0,0 +1,431 @@
+        Returns the significant features based on the q-value cutoff.
+
+    Abstract Methods
+    ----------------
+    _extend_validation(**kwargs)
+        Extends the validation of parameters for the specific method.
+    _perform(**kwargs) -> pd.DataFrame
+        Wrapper that runs the statistical test from kwargs.
+    _run_statistical_test(mat: pd.DataFrame, **kwargs) -> pd.DataFrame
+        Runs the statistical test. The wrapper and the actual method are separated to allow for easier
+        testing and to ensure that all parameters are defined with types and defaults.
+
+    Intended Usage
+    --------------
+        Implement shared methods for two-group analysis.
+        Implement t-test specific methods.
+    result = dea.perform(**settings)  # run once
+    significance = dea.get_significance(cached_result, 0.05)  # run multiple times
+    volcano_plot(cached_result, significance)  # visualize
+    """
+
+    def __init__(self, mat: pd.DataFrame) -> None:
+        """Constructor for the DifferentialExpressionAnalysis class. sets up mat and results.
+
+        Parameters:
+        mat (pd.DataFrame): The input data for the analysis. This should be a DataFrame with the samples as rows and the features as columns.
+        """
+        if mat.empty:
+            raise ValueError(
+                "Input matrix to differential expression analysis is empty."
+            )
+        self.mat = mat
+        self.result: pd.DataFrame = None
+
+    def _validate_input(self, **kwargs) -> None:
+        """Abstract method to validate the input and parameters. This should raise an exception if the input or parameters are invalid
+
+        This function here checks for all parameters required for analysis regardless of the specific method, namely log2_transformed and metadata.
+
+        Parameters:
+        **kwargs (dict): The parameters for the analysis. The keys need to be defined within the allowed_parameters method.
+        """
+        try:
+            self._extend_validation(**kwargs)
+        except TypeError as err:
+            raise TypeError(
+                f"{str(err)}. Accepted keyword arguments to perform are {', '.join(self._allowed_parameters())}."
+            ) from err
+
+        for parameter in kwargs:
+            if parameter not in self._allowed_parameters():
+                raise TypeError(
+                    f"Parameter {parameter} should not be provided for this analysis. Accepted keyword arguments to perform are {', '.join(self._allowed_parameters())}."
+                )
+
+    def _allowed_parameters(self) -> List[str]:
+        """Method returning a list of allowed parameters for the analysis to avoid calling tests with additional parameters."""
+        perform_signature = inspect.signature(self._perform)
+        parameters = list(perform_signature.parameters.keys())
+        return [parameter for parameter in parameters if parameter != "kwargs"]
```

Comment:
can be removed, python does that for you :-)

Reply:
Not if I pass kwargs to _perform though, right? Which I need to do to match the signature of the abstract method? The idea was that by adding this here I enforce kwargs to be empty without relying on every implementation of _perform to contain that check.